File: test-suite/elixir/clean/ssrf.ex
defmodule CleanSsrf do
  @allowed_hosts MapSet.new(["api.example.com", "hooks.example.com"])

  def safe_outbound_url(raw) do
    uri = URI.parse(raw)

    unless uri.scheme == "https" and MapSet.member?(@allowed_hosts, uri.host) do
      raise ArgumentError, "blocked outbound URL"
    end

    raw
  end

  def fetch_req(conn) do
    target = safe_outbound_url(conn.params["url"] || "https://api.example.com/status")
    Req.get!(target)
  end

  def fetch_callback(conn) do
    callback =
      conn
      |> Plug.Conn.get_req_header("x-callback-url")
      |> List.first()
      |> safe_outbound_url()

    HTTPoison.get!(callback)
  end

  def fetch_with_finch(conn) do
    target = safe_outbound_url(conn.params["target"] || "https://api.example.com/status")

    Finch.build(:get, target)
    |> Finch.request(MyApp.Finch)
  end

  def fetch_with_tesla(params) do
    endpoint = safe_outbound_url(Map.get(params, "endpoint") || "https://hooks.example.com")
    Tesla.get(client(), endpoint)
  end

  def fetch_after_inline_validation(conn) do
    raw = conn.params["next_url"] || "https://api.example.com/status"
    uri = URI.parse(raw)

    unless uri.scheme == "https" and MapSet.member?(@allowed_hosts, uri.host) do
      raise ArgumentError, "blocked outbound URL"
    end

    Req.get!(raw)
  end

  defp client do
    Tesla.client([])
  end
end

File: test-suite/rust/clean/ssrf.rs
use std::collections::HashMap;

use axum::extract::Query;
use reqwest::Client;
use url::Url;

fn safe_outbound_url(raw: &str) -> Result<String, &'static str> {
    let parsed = Url::parse(raw).map_err(|_| "invalid url")?;
    if parsed.scheme() != "https" {
        return Err("blocked scheme");
    }
    let allowed_hosts = ["api.example.com", "hooks.example.com"];
    if !allowed_hosts.contains(&parsed.host_str().unwrap_or_default()) {
        return Err("blocked host");
    }
    Ok(parsed.into())
}

async fn fetch_from_query(Query(params): Query<HashMap<String, String>>) -> Result<(), reqwest::Error> {
    let target = safe_outbound_url(params.get("target_url").map(String::as_str).unwrap_or("https://api.example.com"))
        .unwrap_or_else(|_| "https://api.example.com".to_string());
    reqwest::get(target).await?;
    Ok(())
}

async fn fetch_from_header(client: &Client, raw_header: &str) -> Result<(), reqwest::Error> {
    let callback = safe_outbound_url(raw_header).unwrap_or_else(|_| "https://hooks.example.com".to_string());
    client.get(callback).send().await?;
    Ok(())
}

fn fetch_from_config() {
    if let Ok(webhook) = safe_outbound_url("https://hooks.example.com/deploy") {
        let _ = ureq::get(&webhook).call();
    }
}
